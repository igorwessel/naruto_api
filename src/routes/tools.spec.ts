import * as E from 'fp-ts/Either'
import * as TE from 'fp-ts/TaskEither'
import { pipe } from 'fp-ts/function'

import build from '~/app'

const app = build({ logger: false })
const BASE_ROUTE = '/api/v1/tools'

afterAll(() => {
  app.close()
})

it('should get tools', () => {
  return pipe(
    app,
    TE.tryCatchK(
      app =>
        app.inject({
          url: BASE_ROUTE,
        }),
      E.toError
    ),
    TE.map(response => {
      expect(response.statusCode).toBe(200)
      expect(response.json()).toMatchSnapshot()
    })
  )()
})

it.each([
  ['limit', '35'],
  ['offset', '-1'],
])('should get bad request when pass invalid query params', (query, value) => {
  return pipe(
    app,
    TE.tryCatchK(
      app =>
        app.inject({
          url: BASE_ROUTE,
          query: {
            [query]: value,
          },
        }),
      E.toError
    ),
    TE.map(response => {
      expect(response.statusCode).toBe(400)
      expect(response.json()).toMatchObject({ statusCode: 400, error: 'Bad Request' })
    })
  )()
})

it('should paginate tools', () => {
  return pipe(
    app,
    TE.tryCatchK(
      app =>
        app.inject({
          url: BASE_ROUTE,
          query: {
            offset: '1',
          },
        }),
      E.toError
    ),
    TE.map(response => {
      expect(response.statusCode).toBe(200)
      expect(response.json()).toMatchSnapshot()
    })
  )()
})

it('should limit of total tools returned', () => {
  return pipe(
    app,
    TE.tryCatchK(
      app =>
        app.inject({
          url: BASE_ROUTE,
          query: {
            limit: '2',
          },
        }),
      E.toError
    ),
    TE.map(response => {
      expect(response.statusCode).toBe(200)
      expect(response.json()).toHaveLength(2)
    })
  )()
})

it.each([
  ['1', 'id', 1],
  ['a-lenda-de-um-ninja-determinado', 'kebab-case', 1],
])('should get team with %s in format %s', (param, _, id) => {
  return pipe(
    app,
    TE.tryCatchK(
      app =>
        app.inject({
          url: `${BASE_ROUTE}/${param}`,
        }),
      E.toError
    ),
    TE.map(response => {
      expect(response.statusCode).toBe(200)
      expect(response.json()).toMatchObject({ id })
    })
  )()
})

it('should get bad request when pass invalid param', () => {
  return pipe(
    app,
    TE.tryCatchK(
      app =>
        app.inject({
          url: `${BASE_ROUTE}/pascalCase`,
        }),
      E.toError
    ),
    TE.map(response => {
      expect(response.statusCode).toBe(400)
      expect(response.json()).toMatchObject({ statusCode: 400, error: 'Bad Request' })
    })
  )()
})

it('should get error 404 when not found tool', () => {
  return pipe(
    app,
    TE.tryCatchK(
      app =>
        app.inject({
          url: `${BASE_ROUTE}/100000`,
        }),
      E.toError
    ),
    TE.map(response => {
      expect(response.statusCode).toBe(404)
      expect(response.json()).toMatchObject({ statusCode: 404, error: 'Not Found' })
    })
  )()
})

it('should get ninjas from a tools', () => {
  return pipe(
    app,
    TE.tryCatchK(
      app =>
        app.inject({
          url: `${BASE_ROUTE}/2/ninjas`,
        }),
      E.toError
    ),
    TE.map(response => {
      expect(response.statusCode).toBe(200)
      expect(response.json()).toMatchSnapshot()
    })
  )()
})

it('should get error 404 when not found ninjas from a tools', () => {
  return pipe(
    app,
    TE.tryCatchK(
      app =>
        app.inject({
          url: `${BASE_ROUTE}/1/ninjas`,
        }),
      E.toError
    ),
    TE.map(response => {
      expect(response.statusCode).toBe(404)
      expect(response.json()).toMatchObject({ statusCode: 404, error: 'Not Found' })
    })
  )()
})
