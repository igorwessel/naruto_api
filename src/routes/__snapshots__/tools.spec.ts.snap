// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get ninjas from a tools 1`] = `
Array [
  Object {
    "academyGradAge": "10 anos",
    "birthdate": "20 de Marco",
    "bloodType": "A",
    "chuninPromAge": "15 anos",
    "id": 407,
    "name": "Ibiki Morino",
    "ninjaRegistration": "010913",
    "sex": "Masculino",
    "specie": "Humano",
    "status": "Vivo",
  },
  Object {
    "academyGradAge": "12",
    "birthdate": "9 de marco",
    "bloodType": "B",
    "chuninPromAge": "15",
    "id": 1124,
    "name": "Tenten",
    "ninjaRegistration": "012573",
    "sex": "Feminino",
    "specie": "Humano",
    "status": null,
  },
]
`;

exports[`should get tools 1`] = `
Array [
  Object {
    "description": "A Lenda de Um Ninja Determinado foi o primeiro livro de Jiraiya, escrito ao treinar os órfãos de Ame. A história segue as aventuras de um ninja chamado Naruto, nomeado por Jiraiya enquanto ele comia ramen. O personagem principal, Naruto, nunca desiste e promete quebrar a 'maldição' que foi mencionado na história. A 'maldição' mencionada representa os 'ciclos viciosos' de guerra, conflito e ódio, que ocupa o mundo ninja que Nagato uma vez falou sobre com Jiraiya.",
    "games": null,
    "id": 1,
    "kanji": "ド根性忍伝",
    "mangaPanini": null,
    "name": "A Lenda de Um Ninja Determinado",
    "portugues": null,
    "romaji": "Dokonjō Ninden",
    "tvBrasileira": null,
  },
  Object {
    "description": null,
    "games": null,
    "id": 2,
    "kanji": null,
    "mangaPanini": null,
    "name": "Adaga",
    "portugues": null,
    "romaji": null,
    "tvBrasileira": null,
  },
  Object {
    "description": "Uma manopla de metal porosa que está localizada no braço direito de Dosu. Ela é usada para amplificar ou absorver as ondas sonoras provocadas por movimentos de seu braço. A melodia das ondas sonoras provocadas pelo braço podem ser manipuladas e dirigidas com chakra causando mais danos. Elas também podem causar visão borrada, tontura, bem como paralisar o alvo com os jutsu que são provenientes desta arma. Ele também pode agir como um bloqueio contra kunai e shuriken, que por sua vez resulta em mais vibrações que podem ser usadas para a técnica de Dosu.",
    "games": null,
    "id": 3,
    "kanji": "響鳴スピーカー",
    "mangaPanini": null,
    "name": "Alto-Falante Ressonante de Eco",
    "portugues": null,
    "romaji": "Kyōmei Supīkā",
    "tvBrasileira": null,
  },
  Object {
    "description": "Antídotos (解毒薬, Gedokuyaku) são tipos de medicamentos que agem como uma cura para vários venenos. Eles podem ser administrados a uma pessoa infectada por vários meios. Alguns venenos são complexos e, portanto, criar o antídoto vai exigir muita atenção ou então ele poderia matar a vítima; Sasori observou que ao fazer antídotos nem mesmo um erro da linha fina pode ser feito, ou então ele não iria funcionar. No entanto, especialmente treinados, gênios ninja médicos como Tsunade, Shizune ou Sakura Haruno podem realmente analisar e decifrar os muitos elementos diferentes usados ​​em venenos complexos e criar facilmente uma cura. Eles são geralmente um líquido engarrafado, mas pode ser qualquer coisa, desde uma flor ou em pó.",
    "games": null,
    "id": 4,
    "kanji": "解毒薬",
    "mangaPanini": null,
    "name": "Antídoto",
    "portugues": null,
    "romaji": "Gedokuyaku",
    "tvBrasileira": null,
  },
  Object {
    "description": "Um Arco & Flecha (弓矢, Yumiya) é uma arma de condução usado para disparar flechas em um alvo distante. Esta arma consiste em um ou mais fios em ambas as extremidades ligadas a lados opostos de um sólido, parte flexível curva ou um conjunto de várias peças de bloqueio (compostos Arcos) que pode consistir de madeira, osso, fibra de carbono ou outra materiais. Jutsu baseado no uso de arco e flecha são chamados Kyūjutsu (弓术, literalmente significa 'Técnica Arco'). Kidōmaru por sua vez cria arcos e usando suas teias de aranha pegajosa de Ouro que lhe permite criar um arco poderoso que pode fazer melhor uso usando ambos os braços. Além disso, Uchiha Sasuke usado uma besta em seu Susanoo que é melhorada pelo tamanho e poder destrutivo. Depois de colocar o chakra das Bijū em todo o corpo de seu Susanoo, Sasuke pode formar um arco feito de puro raio que lhe permite atirar uma flecha de energia altamente destrutiva, que declarou como seu ataque mais poderoso. No Anime, Tenten usa uma espécie de arco com quatro flechas carregadas de uma vez. Para firmar este arco deve cair mecanismo de barreiras em vez de simplesmente liberar a corda. Ela usou -o para atacar à distância quando Gennō estava voando em sua ave gigante.",
    "games": null,
    "id": 5,
    "kanji": "弓矢",
    "mangaPanini": null,
    "name": "Arco & Flecha",
    "portugues": null,
    "romaji": "Yumiya",
    "tvBrasileira": null,
  },
  Object {
    "description": "Areia (砂, Suna), é um material versátil, mais frequentemente utilizado pelos shinobi proveniente de Sunagakure para diversos fins, incluindo várias funções complementares que vão desde a ocultação de disfarçar. No entanto, a própria ferramenta ganhou destaque nas mãos do jinchūriki do Shukaku, que adquiriu a habilidade de manipular a areia através da besta, com o mais famoso entre os quais está Gaara.",
    "games": null,
    "id": 6,
    "kanji": "砂",
    "mangaPanini": null,
    "name": "Areia",
    "portugues": null,
    "romaji": "Suna",
    "tvBrasileira": null,
  },
  Object {
    "description": "Areia de Ferro (砂鉄, Satetsu) é um conjunto de grânulos de ferro que podem ser usados como uma ferramenta através de uma habilidade única de converter seu chakra em forças magnéticas.",
    "games": null,
    "id": 7,
    "kanji": "砂鉄",
    "mangaPanini": null,
    "name": "Areia de Ferro",
    "portugues": null,
    "romaji": "Satetsu",
    "tvBrasileira": "Areia Ferruginosa",
  },
  Object {
    "description": "São argolas semelhantes a chakrans, que são lançadas no intuito de capturar um oponente. Quando o usuário lança essa ferramenta ainda fechada, elas se abrem perto do alvo, com a parte simples passando pela dupla, até que fecham com a mesma mecânica e restringem o inimigo.",
    "games": null,
    "id": 8,
    "kanji": null,
    "mangaPanini": null,
    "name": "Argolas de Captura",
    "portugues": null,
    "romaji": null,
    "tvBrasileira": null,
  },
  Object {
    "description": "A Armadura sugando chakra.",
    "games": null,
    "id": 9,
    "kanji": "無限甲蓋",
    "mangaPanini": null,
    "name": "Armadura Infinita",
    "portugues": null,
    "romaji": "Mugen Kōgai",
    "tvBrasileira": null,
  },
  Object {
    "description": "Durante a era antes das vilas ocultas, os shinobi usavam uma armadura que era semelhante ao de um samurai. Ela proporcionava uma defesa maior do que o vestuário simples que os shinobi usariam em missões que não seriam para batalhas decisivas devido a ser feito de placas de armadura pesadas. Os shinobi de Konoha continuaram a usar a armadura até pelo menos a Primeira Guerra Mundial Shinobi, quando na segunda a transição para coletes começou.",
    "games": null,
    "id": 10,
    "kanji": null,
    "mangaPanini": null,
    "name": "Armadura de Batalha Shinobi",
    "portugues": null,
    "romaji": null,
    "tvBrasileira": null,
  },
  Object {
    "description": "A Armadura de Chakra (チャクラの鎧, Chakura no Yoroi) é um tipo único de armadura que drena chakra de ninjutsu e genjutsu, enquanto ao mesmo tempo amplifica o chakra do usuário. A armadura com o protótipo de cor preta foi usada por Dotō Kazahana.",
    "games": null,
    "id": 11,
    "kanji": "チャクラの鎧",
    "mangaPanini": null,
    "name": "Armadura de Chakra",
    "portugues": null,
    "romaji": "Chakura no Yoroi",
    "tvBrasileira": null,
  },
  Object {
    "description": "A Armadura de Vapor (蒸気の鎧, Jōki no Yoroi) é uma armadura de metal usada por Han, como instrumento de filtragem do vapor usado em suas técnicas.",
    "games": null,
    "id": 12,
    "kanji": "蒸気の鎧",
    "mangaPanini": null,
    "name": "Armadura de Vapor",
    "portugues": null,
    "romaji": "Jōki no Yoroi",
    "tvBrasileira": null,
  },
  Object {
    "description": "Esta é uma das ferramentas utilizadas pelos ninjas médicos de Konohagakure. Assemelha-se a um fuzil de precisão rudimentar, que é usado para atirar projéteis, como tiros de injeção para desacordar o alvo. É constituída por um tubo de metal com aproximadamente o tamanho de uma régua com uma abertura em ambas as extremidades, como um tubo de descarga e está equipado com um espaço na parte superior e um laser de vista sobre a parte inferior. A arma é projetada para lançar seringas a longa distância, depois que o usuário pressionar a arma com a palma da mão, o dardo é atirado.",
    "games": null,
    "id": 13,
    "kanji": null,
    "mangaPanini": null,
    "name": "Atirador de Dardos de Injeção",
    "portugues": null,
    "romaji": null,
    "tvBrasileira": null,
  },
  Object {
    "description": "O Bō (棒, Literalmente significa: Bastão) é uma arma que é implementada na arte marcial japonesa e é a base do bōjutsu. Esta arma consiste em longa vara de madeira pura.",
    "games": null,
    "id": 14,
    "kanji": "棒",
    "mangaPanini": null,
    "name": "Bō",
    "portugues": null,
    "romaji": "Bō",
    "tvBrasileira": null,
  },
  Object {
    "description": "A balista é um lançador móvel de flechas maior que uma pessoa comum que é usado para acertar alvos a longa distância.",
    "games": null,
    "id": 15,
    "kanji": null,
    "mangaPanini": null,
    "name": "Balista",
    "portugues": null,
    "romaji": null,
    "tvBrasileira": null,
  },
]
`;

exports[`should paginate tools 1`] = `
Array [
  Object {
    "description": null,
    "games": null,
    "id": 2,
    "kanji": null,
    "mangaPanini": null,
    "name": "Adaga",
    "portugues": null,
    "romaji": null,
    "tvBrasileira": null,
  },
  Object {
    "description": "Uma manopla de metal porosa que está localizada no braço direito de Dosu. Ela é usada para amplificar ou absorver as ondas sonoras provocadas por movimentos de seu braço. A melodia das ondas sonoras provocadas pelo braço podem ser manipuladas e dirigidas com chakra causando mais danos. Elas também podem causar visão borrada, tontura, bem como paralisar o alvo com os jutsu que são provenientes desta arma. Ele também pode agir como um bloqueio contra kunai e shuriken, que por sua vez resulta em mais vibrações que podem ser usadas para a técnica de Dosu.",
    "games": null,
    "id": 3,
    "kanji": "響鳴スピーカー",
    "mangaPanini": null,
    "name": "Alto-Falante Ressonante de Eco",
    "portugues": null,
    "romaji": "Kyōmei Supīkā",
    "tvBrasileira": null,
  },
  Object {
    "description": "Antídotos (解毒薬, Gedokuyaku) são tipos de medicamentos que agem como uma cura para vários venenos. Eles podem ser administrados a uma pessoa infectada por vários meios. Alguns venenos são complexos e, portanto, criar o antídoto vai exigir muita atenção ou então ele poderia matar a vítima; Sasori observou que ao fazer antídotos nem mesmo um erro da linha fina pode ser feito, ou então ele não iria funcionar. No entanto, especialmente treinados, gênios ninja médicos como Tsunade, Shizune ou Sakura Haruno podem realmente analisar e decifrar os muitos elementos diferentes usados ​​em venenos complexos e criar facilmente uma cura. Eles são geralmente um líquido engarrafado, mas pode ser qualquer coisa, desde uma flor ou em pó.",
    "games": null,
    "id": 4,
    "kanji": "解毒薬",
    "mangaPanini": null,
    "name": "Antídoto",
    "portugues": null,
    "romaji": "Gedokuyaku",
    "tvBrasileira": null,
  },
  Object {
    "description": "Um Arco & Flecha (弓矢, Yumiya) é uma arma de condução usado para disparar flechas em um alvo distante. Esta arma consiste em um ou mais fios em ambas as extremidades ligadas a lados opostos de um sólido, parte flexível curva ou um conjunto de várias peças de bloqueio (compostos Arcos) que pode consistir de madeira, osso, fibra de carbono ou outra materiais. Jutsu baseado no uso de arco e flecha são chamados Kyūjutsu (弓术, literalmente significa 'Técnica Arco'). Kidōmaru por sua vez cria arcos e usando suas teias de aranha pegajosa de Ouro que lhe permite criar um arco poderoso que pode fazer melhor uso usando ambos os braços. Além disso, Uchiha Sasuke usado uma besta em seu Susanoo que é melhorada pelo tamanho e poder destrutivo. Depois de colocar o chakra das Bijū em todo o corpo de seu Susanoo, Sasuke pode formar um arco feito de puro raio que lhe permite atirar uma flecha de energia altamente destrutiva, que declarou como seu ataque mais poderoso. No Anime, Tenten usa uma espécie de arco com quatro flechas carregadas de uma vez. Para firmar este arco deve cair mecanismo de barreiras em vez de simplesmente liberar a corda. Ela usou -o para atacar à distância quando Gennō estava voando em sua ave gigante.",
    "games": null,
    "id": 5,
    "kanji": "弓矢",
    "mangaPanini": null,
    "name": "Arco & Flecha",
    "portugues": null,
    "romaji": "Yumiya",
    "tvBrasileira": null,
  },
  Object {
    "description": "Areia (砂, Suna), é um material versátil, mais frequentemente utilizado pelos shinobi proveniente de Sunagakure para diversos fins, incluindo várias funções complementares que vão desde a ocultação de disfarçar. No entanto, a própria ferramenta ganhou destaque nas mãos do jinchūriki do Shukaku, que adquiriu a habilidade de manipular a areia através da besta, com o mais famoso entre os quais está Gaara.",
    "games": null,
    "id": 6,
    "kanji": "砂",
    "mangaPanini": null,
    "name": "Areia",
    "portugues": null,
    "romaji": "Suna",
    "tvBrasileira": null,
  },
  Object {
    "description": "Areia de Ferro (砂鉄, Satetsu) é um conjunto de grânulos de ferro que podem ser usados como uma ferramenta através de uma habilidade única de converter seu chakra em forças magnéticas.",
    "games": null,
    "id": 7,
    "kanji": "砂鉄",
    "mangaPanini": null,
    "name": "Areia de Ferro",
    "portugues": null,
    "romaji": "Satetsu",
    "tvBrasileira": "Areia Ferruginosa",
  },
  Object {
    "description": "São argolas semelhantes a chakrans, que são lançadas no intuito de capturar um oponente. Quando o usuário lança essa ferramenta ainda fechada, elas se abrem perto do alvo, com a parte simples passando pela dupla, até que fecham com a mesma mecânica e restringem o inimigo.",
    "games": null,
    "id": 8,
    "kanji": null,
    "mangaPanini": null,
    "name": "Argolas de Captura",
    "portugues": null,
    "romaji": null,
    "tvBrasileira": null,
  },
  Object {
    "description": "A Armadura sugando chakra.",
    "games": null,
    "id": 9,
    "kanji": "無限甲蓋",
    "mangaPanini": null,
    "name": "Armadura Infinita",
    "portugues": null,
    "romaji": "Mugen Kōgai",
    "tvBrasileira": null,
  },
  Object {
    "description": "Durante a era antes das vilas ocultas, os shinobi usavam uma armadura que era semelhante ao de um samurai. Ela proporcionava uma defesa maior do que o vestuário simples que os shinobi usariam em missões que não seriam para batalhas decisivas devido a ser feito de placas de armadura pesadas. Os shinobi de Konoha continuaram a usar a armadura até pelo menos a Primeira Guerra Mundial Shinobi, quando na segunda a transição para coletes começou.",
    "games": null,
    "id": 10,
    "kanji": null,
    "mangaPanini": null,
    "name": "Armadura de Batalha Shinobi",
    "portugues": null,
    "romaji": null,
    "tvBrasileira": null,
  },
  Object {
    "description": "A Armadura de Chakra (チャクラの鎧, Chakura no Yoroi) é um tipo único de armadura que drena chakra de ninjutsu e genjutsu, enquanto ao mesmo tempo amplifica o chakra do usuário. A armadura com o protótipo de cor preta foi usada por Dotō Kazahana.",
    "games": null,
    "id": 11,
    "kanji": "チャクラの鎧",
    "mangaPanini": null,
    "name": "Armadura de Chakra",
    "portugues": null,
    "romaji": "Chakura no Yoroi",
    "tvBrasileira": null,
  },
  Object {
    "description": "A Armadura de Vapor (蒸気の鎧, Jōki no Yoroi) é uma armadura de metal usada por Han, como instrumento de filtragem do vapor usado em suas técnicas.",
    "games": null,
    "id": 12,
    "kanji": "蒸気の鎧",
    "mangaPanini": null,
    "name": "Armadura de Vapor",
    "portugues": null,
    "romaji": "Jōki no Yoroi",
    "tvBrasileira": null,
  },
  Object {
    "description": "Esta é uma das ferramentas utilizadas pelos ninjas médicos de Konohagakure. Assemelha-se a um fuzil de precisão rudimentar, que é usado para atirar projéteis, como tiros de injeção para desacordar o alvo. É constituída por um tubo de metal com aproximadamente o tamanho de uma régua com uma abertura em ambas as extremidades, como um tubo de descarga e está equipado com um espaço na parte superior e um laser de vista sobre a parte inferior. A arma é projetada para lançar seringas a longa distância, depois que o usuário pressionar a arma com a palma da mão, o dardo é atirado.",
    "games": null,
    "id": 13,
    "kanji": null,
    "mangaPanini": null,
    "name": "Atirador de Dardos de Injeção",
    "portugues": null,
    "romaji": null,
    "tvBrasileira": null,
  },
  Object {
    "description": "O Bō (棒, Literalmente significa: Bastão) é uma arma que é implementada na arte marcial japonesa e é a base do bōjutsu. Esta arma consiste em longa vara de madeira pura.",
    "games": null,
    "id": 14,
    "kanji": "棒",
    "mangaPanini": null,
    "name": "Bō",
    "portugues": null,
    "romaji": "Bō",
    "tvBrasileira": null,
  },
  Object {
    "description": "A balista é um lançador móvel de flechas maior que uma pessoa comum que é usado para acertar alvos a longa distância.",
    "games": null,
    "id": 15,
    "kanji": null,
    "mangaPanini": null,
    "name": "Balista",
    "portugues": null,
    "romaji": null,
    "tvBrasileira": null,
  },
  Object {
    "description": "O Bashōsen (芭蕉扇; Literalmente significa 'Leque do Palmo de Banana') é um grande leque capaz de gerar todas as cinco naturezas de chakra elementais básicas. Para fazer isso, o usuário só precisa dizer um comando mental, que produz quantidades substanciais do elemento desejado quando o leque é ondulado.",
    "games": null,
    "id": 16,
    "kanji": "芭蕉扇",
    "mangaPanini": null,
    "name": "Bashōsen",
    "portugues": null,
    "romaji": "Bashōsen",
    "tvBrasileira": null,
  },
]
`;
